experimental/DrinfeldHeckeAlgebras/src/DrinfeldHeckeAlgebra.jl

```diff
@@ -0,0 +1,669 @@
+parent(a::DrinfeldHeckeAlgebraElem) = a.parent
+is_domain_type(::Type{DrinfeldHeckeAlgebraElem}) = false
+is_exact_type(::Type{DrinfeldHeckeAlgebraElem}) = true
+Base.hash(a::DrinfeldHeckeAlgebraElem, h::UInt) = Base.hash(a.element, h)
+deepcopy_internal(a::DrinfeldHeckeAlgebraElem, dict::IdDict) = DrinfeldHeckeAlgebraElem(a.parent, a.element)
+
+#######################################
+# Basic manipulation of rings and elements
+#######################################
+
+zero(A::DrinfeldHeckeAlgebra) = A()
+one(A::DrinfeldHeckeAlgebra) = A(1)
+iszero(a::DrinfeldHeckeAlgebraElem) = iszero(a.element)
+isone(a::DrinfeldHeckeAlgebraElem) = isone(a.element)
+
+#######################################
+# Canonicalisation
+#######################################
+
+canonical_unit(a::DrinfeldHeckeAlgebraElem) = one(a.parent)
+
+#######################################
+# Arithmetic Operations
+#######################################
+
+-(a::DrinfeldHeckeAlgebraElem) = DrinfeldHeckeAlgebraElem(a.parent, -a.element)
+
++(a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem) = DrinfeldHeckeAlgebraElem(a.parent, a.element + b.element)
+-(a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem) = DrinfeldHeckeAlgebraElem(a.parent, a.element - b.element)
+*(a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem) = multiply(a,b)
+
+==(a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem) = a.element == b.element
+isequal(a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem) = a == b
+
+function ^(a::DrinfeldHeckeAlgebraElem, e::Int)
+  if e == 0 return one(a.parent) end
+  if e == 1 return a end
+  
+  return multiply(a,a^(e - 1))
+end
+
+#######################################
+# Type coercion
+#######################################
+
+*(a::Union{RingElem, MatrixGroupElem}, b::DrinfeldHeckeAlgebraElem) = b.parent(a) * b.parent(b)
+*(a::DrinfeldHeckeAlgebraElem, b::Union{RingElem, MatrixGroupElem}) = a.parent(a) * a.parent(b)
++(a::Union{RingElem, MatrixGroupElem}, b::DrinfeldHeckeAlgebraElem) = b.parent(a) + b.parent(b)
++(a::DrinfeldHeckeAlgebraElem, b::Union{RingElem, MatrixGroupElem}) = a.parent(a) + a.parent(b)
+-(a::Union{RingElem, MatrixGroupElem}, b::DrinfeldHeckeAlgebraElem) = b.parent(a) - b.parent(b)
+-(a::DrinfeldHeckeAlgebraElem, b::Union{RingElem, MatrixGroupElem}) = a.parent(a) - a.parent(b)
+
+#######################################
+# Unsafe operators
+#######################################
+
+function zero!(a::DrinfeldHeckeAlgebraElem)
+  a = zero(a.parent)
+  return a
+end
+
+function mul!(c::DrinfeldHeckeAlgebraElem, a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem)
+  c = a * b
+  return c
+end
+
+function add!(c::DrinfeldHeckeAlgebraElem, a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem)
+  c = a + b
+  return c
+end
+
+function addeq!(a::DrinfeldHeckeAlgebraElem, b::DrinfeldHeckeAlgebraElem)
+  a = a + b
+  return a
+end
```

Comment:
```suggestion
```
these at the moment are not unsafe, and thus are no better than the generic fallback implementation. Thus we can just leave them out here